x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/AutoFollowCoordinator.java

```diff
@@ -0,0 +1,261 @@
+                List<String> followedIndices = autoFollowMetadata.getFollowedLeaderIndexUUIDS().get(clusterAlias);
+
+                clusterStateApiCall(remoteClient, (remoteClusterState, e) -> {
+                    if (remoteClusterState != null) {
+                        handleClusterAlias(clusterAlias, autoFollowPattern, followedIndices, remoteClusterState);
+                    } else {
+                        finalise(e);
+                    }
+                });
+            }
+        }
+
+        private void handleClusterAlias(String clusterAlias, AutoFollowMetadata.AutoFollowPattern autoFollowPattern,
+                                List<String> followedIndexUUIDS, ClusterState remoteClusterState) {
+            List<IndexMetaData> leaderIndicesToFollow = new ArrayList<>();
+            String[] patterns = autoFollowPattern.getLeaderIndexPatterns().toArray(new String[0]);
+            for (IndexMetaData indexMetaData : remoteClusterState.getMetaData()) {
+                if (Regex.simpleMatch(patterns, indexMetaData.getIndex().getName())) {
+                    if (followedIndexUUIDS.contains(indexMetaData.getIndex().getUUID()) == false) {
+                        leaderIndicesToFollow.add(indexMetaData);
+                    }
+                }
+            }
+            if (leaderIndicesToFollow.isEmpty()) {
+                finalise(null);
+            } else {
+                AtomicInteger numRequests = new AtomicInteger();
+                for (IndexMetaData indexToFollow : leaderIndicesToFollow) {
+                    String leaderIndexName = indexToFollow.getIndex().getName();
+                    String followIndexName = leaderIndexName;
+                    if (autoFollowPattern.getFollowIndexPattern() != null) {
+                        followIndexName = autoFollowPattern.getFollowIndexPattern().replace("{{leader_index}}", leaderIndexName);
+                    }
+
+                    String leaderIndexNameWithClusterAliasPrefix = clusterAlias.equals("_local_") ? leaderIndexName :
+                        clusterAlias + ":" + leaderIndexName;
+                    FollowIndexAction.Request followRequest =
+                        new FollowIndexAction.Request(leaderIndexNameWithClusterAliasPrefix, followIndexName,
+                            autoFollowPattern.getMaxBatchOperationCount(), autoFollowPattern.getMaxConcurrentReadBatches(),
+                            autoFollowPattern.getMaxOperationSizeInBytes(), autoFollowPattern.getMaxConcurrentWriteBatches(),
+                            autoFollowPattern.getMaxWriteBufferSize(), autoFollowPattern.getRetryTimeout(),
+                            autoFollowPattern.getIdleShardRetryDelay());
+
+                    // This runs on the elected master node, so we can update cluster state here:
+                    Consumer<Exception> handler = followError -> {
```

Comment:
This is difficult to read because it breaks the usual linear flow of reading code. So we are defining a callback, and then within that defining another callback and then it's really the `updateAutoMetadata` that executes first, and then there's another callback defined. I think it might be easier if we broke it down into `onSuccess/onFailure` instead of the `Consumer<Exception>` so that we can see the logic separately and think of it as "this is what happens if there is success, and this is what happens if there is failure".

Reply:
Because really what we are after here is:
 - try to create the follower index and follow the leader
 - if that succeeds, update the auto-follow metadata, then do some accounting
 - if that fails, do some accounting and track the failure

it would be nice if the code read a little closer to that.